File: src/main/java/no/nav/veilarboppfolging/controller/AdminController.java
package no.nav.veilarboppfolging.controller;

import no.nav.common.auth.context.AuthContextHolder;
import no.nav.common.job.JobRunner;
import no.nav.common.types.identer.AktorId;
import no.nav.common.types.identer.NavIdent;
import no.nav.veilarboppfolging.ForbiddenException;
import no.nav.veilarboppfolging.controller.response.Veilarbportefoljeinfo;
import no.nav.veilarboppfolging.controller.v2.request.RepubliserVeilederRequest;
import no.nav.veilarboppfolging.domain.AvsluttOppfolgingsperiodePayload;
import no.nav.veilarboppfolging.domain.AvsluttResultat;
import no.nav.veilarboppfolging.domain.RepubliserOppfolgingsperioderRequest;
import no.nav.veilarboppfolging.domain.AvsluttPayload;
import no.nav.veilarboppfolging.kandidatForUtmelding.KandidatForUtmeldingService;
import no.nav.veilarboppfolging.oppfolgingsbruker.VeilederRegistrant;
import no.nav.veilarboppfolging.oppfolgingsbruker.utgang.AdminAvregistrering;
import no.nav.veilarboppfolging.repository.OppfolgingsPeriodeRepository;
import no.nav.veilarboppfolging.repository.VeilederTilordningerRepository;
import no.nav.veilarboppfolging.repository.entity.ManuellStatusEntity;
import no.nav.veilarboppfolging.repository.entity.OppfolgingsperiodeEntity;
import no.nav.veilarboppfolging.repository.entity.VeilederTilordningEntity;
import no.nav.veilarboppfolging.service.*;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.web.bind.annotation.*;

import java.time.ZonedDateTime;
import java.util.List;
import java.util.Optional;

@RestController
@RequestMapping("/api/admin")
public class AdminController {

    public static final String POAO_ADMIN = "poao-admin";
    private final AuthService authService;
    private final AuthContextHolder authContextHolder;
    private final KafkaRepubliseringService kafkaRepubliseringService;
    private final VeilederTilordningerRepository veilederTilordningerRepository;
    private final ManuellStatusService manuellStatusService;
    private final OppfolgingsPeriodeRepository oppfolgingsPeriodeRepository;
    private final OppfolgingService oppfolgingService;
    private final AvsluttOppfolgingService avsluttOppfolgingService;
    private final KandidatForUtmeldingService kandidatForUtmeldingService;

    private final Logger log = LoggerFactory.getLogger(this.getClass());

    @Autowired
    public AdminController(
            AuthService authService,
            AuthContextHolder authContextHolder,
            KafkaRepubliseringService kafkaRepubliseringService,
            VeilederTilordningerRepository veilederTilordningerRepository,
            ManuellStatusService manuellStatusService,
            OppfolgingsPeriodeRepository oppfolgingsPeriodeRepository,
            OppfolgingService oppfolgingService,
            AvsluttOppfolgingService avsluttOppfolgingService,
            KandidatForUtmeldingService kandidatForUtmeldingService
    ) {
        this.authService = authService;
        this.authContextHolder = authContextHolder;
        this.kafkaRepubliseringService = kafkaRepubliseringService;
        this.veilederTilordningerRepository = veilederTilordningerRepository;
        this.manuellStatusService = manuellStatusService;
        this.oppfolgingsPeriodeRepository = oppfolgingsPeriodeRepository;
        this.oppfolgingService = oppfolgingService;
        this.avsluttOppfolgingService = avsluttOppfolgingService;
        this.kandidatForUtmeldingService = kandidatForUtmeldingService;
    }

    @PostMapping("/republiser/oppfolgingsperioder")
    public String republiserOppfolgingsperioder(@RequestBody(required = false) RepubliserOppfolgingsperioderRequest request) {
        sjekkTilgangTilAdmin();

        if (request != null && request.aktorId != null) {
            return JobRunner.runAsync("republiser-oppfolgingsperioder-for-bruker", () -> kafkaRepubliseringService.republiserOppfolgingsperiodeForBruker(request.aktorId));
        }

        return JobRunner.runAsync("republiser-oppfolgingsperioder", kafkaRepubliseringService::republiserOppfolgingsperioder);
    }

    @PostMapping("/republiser/tilordnet-veileder")
    public String republiserTilordnetVeileder() {
        sjekkTilgangTilAdmin();
        return JobRunner.runAsync("republiser-tilordnet-veileder", kafkaRepubliseringService::republiserTilordnetVeileder);
    }

    @PostMapping("/republiser/tilordnet-veileder/utvalg")
    public String republiserTilordnetVeileder(@RequestBody RepubliserVeilederRequest republiserVeilederRequest) {
        sjekkTilgangTilAdmin();
        return JobRunner.runAsync(
                "republiser-tilordnet-veileder-gitte-aktorider",
                () -> kafkaRepubliseringService.republiserTilordnetVeileder(republiserVeilederRequest.aktorIder())
        );
    }

    @PostMapping("/republiser/kvp-perioder")
    public String republiserKvpPerioder() {
        sjekkTilgangTilAdmin();
        return JobRunner.runAsync("republiser-kvp-perioder", kafkaRepubliseringService::republiserKvpPerioder);
    }

    @GetMapping("/hentVeilarbinfo/bruker")
    public Veilarbportefoljeinfo hentVeilarbportefoljeinfo(@RequestParam AktorId aktorId) {
        authService.skalVereSystemBruker();
        Optional<VeilederTilordningEntity> tilordningEntity = veilederTilordningerRepository.hentTilordnetVeileder(aktorId);
        boolean erManuell = manuellStatusService.hentManuellStatus(aktorId).map(ManuellStatusEntity::getManuell).orElse(false);
        ZonedDateTime startDato = oppfolgingsPeriodeRepository.hentGjeldendeOppfolgingsperiode(aktorId).map(OppfolgingsperiodeEntity::getStartDato).orElse(null);

        return new Veilarbportefoljeinfo(
                aktorId,
                tilordningEntity.map(VeilederTilordningEntity::getVeilederId).map(NavIdent::of).orElse(null),
                tilordningEntity.map(VeilederTilordningEntity::getOppfolging).orElse(false),
                tilordningEntity.map(VeilederTilordningEntity::getNyForVeileder).orElse(false),
                erManuell,
                startDato,
                tilordningEntity.map(VeilederTilordningEntity::getSistTilordnet).orElse(null)
        );
    }

    @PostMapping("/avsluttBrukere")
    public AvsluttResultat batchAvsluttBrukere(@RequestBody AvsluttPayload brukereSomSkalAvsluttes) {
        sjekkTilgangTilAdmin();
        var innloggetBruker = authService.getInnloggetVeilederIdent();
        log.info("Skal avslutte oppfølging for {} brukere", brukereSomSkalAvsluttes.aktorIds.size());

        var resultat = brukereSomSkalAvsluttes.aktorIds
                .stream()
                .map(aktorId -> {
                    try {
                        avsluttOppfolgingService.adminAvsluttOppfolgingForBruker(
                                new AdminAvregistrering(
                                        AktorId.of(aktorId),
                                        new VeilederRegistrant(new NavIdent(innloggetBruker)),
                                        brukereSomSkalAvsluttes.begrunnelse,
                                        null
                                )
                        );
                        return true;
                    } catch (Exception e) {
                        log.warn("Kunne ikke avslutte oppfølging: {}", e.getMessage());
                        return false;
                    }
                }).toList();

        var avsluttedeBrukere = resultat.stream().filter(it -> it).toList().size();
        var ikkeAvsluttedeBrukere = resultat.stream().filter(it -> !it).toList().size();

        log.info("Avsluttet oppfølging for {} brukere", avsluttedeBrukere);
        log.info("Kunne ikke avslutte oppfølging for {} brukere", ikkeAvsluttedeBrukere);

        return new AvsluttResultat(avsluttedeBrukere, ikkeAvsluttedeBrukere);
    }

    @PostMapping("/avsluttOppfolgingsperiode")
    public boolean avsluttOppfolgingsperiode(@RequestBody AvsluttOppfolgingsperiodePayload oppfolgingsperiodeSomSkalAvsluttes) {
        sjekkTilgangTilAdmin();
        var innloggetBruker = authService.getInnloggetVeilederIdent();

        try {
            avsluttOppfolgingService.adminAvsluttSpesifikkOppfolgingsperiode(
                    AktorId.of(oppfolgingsperiodeSomSkalAvsluttes.getAktorId()),
                    innloggetBruker,
                    oppfolgingsperiodeSomSkalAvsluttes.getBegrunnelse(),
                    oppfolgingsperiodeSomSkalAvsluttes.getOppfolgingsperiodeUuid());
            return true;
        } catch (Exception e) {
            log.warn("Kunne ikke avslutte oppfølgingsperiode: {}", e.getMessage());
            return false;
        }
    }

    private void sjekkTilgangTilAdmin() {
        if (!authService.erInternBruker()) throw new ForbiddenException("Må være internbruker");
        authService.sjekkAtApplikasjonErIAllowList(List.of(POAO_ADMIN));
    }

}

File: src/main/kotlin/no/nav/veilarboppfolging/controller/admin/v1/AdminController.kt
package no.nav.veilarboppfolging.controller.admin.v1

import kotlin.jvm.optionals.getOrNull
import no.nav.common.job.JobRunner
import no.nav.common.types.identer.AktorId
import no.nav.common.types.identer.NavIdent
import no.nav.veilarboppfolging.ForbiddenException
import no.nav.veilarboppfolging.controller.response.Veilarbportefoljeinfo
import no.nav.veilarboppfolging.controller.v2.request.RepubliserVeilederRequest
import no.nav.veilarboppfolging.domain.AvsluttOppfolgingsperiodePayload
import no.nav.veilarboppfolging.domain.RepubliserOppfolgingsperioderRequest
import no.nav.veilarboppfolging.oppfolgingsbruker.VeilederRegistrant
import no.nav.veilarboppfolging.oppfolgingsbruker.utgang.AdminAvregistrering
import no.nav.veilarboppfolging.repository.OppfolgingsPeriodeRepository
import no.nav.veilarboppfolging.repository.VeilederTilordningerRepository
import no.nav.veilarboppfolging.service.AuthService
import no.nav.veilarboppfolging.service.AvsluttOppfolgingService
import no.nav.veilarboppfolging.service.KafkaRepubliseringService
import no.nav.veilarboppfolging.service.ManuellStatusService
import no.nav.veilarboppfolging.service.RepubliserOppfolgingshendelseService
import org.slf4j.LoggerFactory
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RequestParam
import org.springframework.web.bind.annotation.RestController

const val POAO_ADMIN = "poao-admin"

@RestController
@RequestMapping("/api/admin")
class AdminController(
    private val authService: AuthService,
    private val kafkaRepubliseringService: KafkaRepubliseringService,
    private val veilederTilordningerRepository: VeilederTilordningerRepository,
    private val manuellStatusService: ManuellStatusService,
    private val oppfolgingsPeriodeRepository: OppfolgingsPeriodeRepository,
    private val avsluttOppfolgingService: AvsluttOppfolgingService,
    private val republiserOppfolgingshendelseService: RepubliserOppfolgingshendelseService,
) {
    private val log = LoggerFactory.getLogger(this::class.java)

    @PostMapping("/republiser/oppfolgingsperioder")
    fun republiserOppfolgingsperioder(@RequestBody(required = false) request: RepubliserOppfolgingsperioderRequest?): String {
        sjekkTilgangTilAdmin()

        if (request != null) {
            return JobRunner.runAsync(
                "republiser-oppfolgingsperioder-for-bruker"
            ) { kafkaRepubliseringService.republiserOppfolgingsperiodeForBruker(request.aktorId) }
        }

        return JobRunner.runAsync(
            "republiser-oppfolgingsperioder"
        ) { kafkaRepubliseringService.republiserOppfolgingsperioder() }
    }

    @PostMapping("/republiser/tilordnet-veileder")
    fun republiserTilordnetVeileder(): String {
        sjekkTilgangTilAdmin()
        return JobRunner.runAsync(
            "republiser-tilordnet-veileder"
        ) { kafkaRepubliseringService.republiserTilordnetVeileder() }
    }

    @PostMapping("/republiser/tilordnet-veileder/utvalg")
    fun republiserTilordnetVeileder(@RequestBody republiserVeilederRequest: RepubliserVeilederRequest): String {
        sjekkTilgangTilAdmin()
        return JobRunner.runAsync(
            "republiser-tilordnet-veileder-gitte-aktorider"
        ) { kafkaRepubliseringService.republiserTilordnetVeileder(republiserVeilederRequest.aktorIder) }
    }

    @PostMapping("/republiser/kvp-perioder")
    fun republiserKvpPerioder(): String {
        sjekkTilgangTilAdmin()
        return JobRunner.runAsync("republiser-kvp-perioder") { kafkaRepubliseringService.republiserKvpPerioder() }
    }

    @GetMapping("/hentVeilarbinfo/bruker")
    fun hentVeilarbportefoljeinfo(@RequestParam aktorId: AktorId): Veilarbportefoljeinfo {
        authService.skalVereSystemBruker()
        val tilordningEntity = veilederTilordningerRepository.hentTilordnetVeileder(aktorId).getOrNull()
        val erManuell = manuellStatusService.hentManuellStatus(aktorId).getOrNull()?.manuell ?: false
        val startDato = oppfolgingsPeriodeRepository.hentGjeldendeOppfolgingsperiode(aktorId).getOrNull()?.startDato

        return Veilarbportefoljeinfo(
            aktorId = aktorId,
            veilederId = tilordningEntity?.veilederId?.let { NavIdent.of(it) },
            erUnderOppfolging = tilordningEntity?.oppfolging ?: false,
            nyForVeileder = tilordningEntity?.nyForVeileder ?: false,
            erManuell = erManuell,
            startDato = startDato,
            tilordnetTidspunkt = tilordningEntity?.sistTilordnet,
        )
    }

    @PostMapping("/avsluttBrukere")
    fun batchAvsluttBrukere(@RequestBody brukereSomSkalAvsluttes: AvsluttPayload): AvsluttResultat {
        sjekkTilgangTilAdmin()
        val innloggetBruker = authService.getInnloggetVeilederIdent()
        log.info("Skal avslutte oppfølging for {} brukere", brukereSomSkalAvsluttes.aktorIds.size)

        val resultat = brukereSomSkalAvsluttes.aktorIds.map { aktorId ->
            runCatching {
                avsluttOppfolgingService.adminAvsluttOppfolgingForBruker(
                    AdminAvregistrering(
                        AktorId.of(aktorId),
                        VeilederRegistrant(NavIdent(innloggetBruker)),
                        brukereSomSkalAvsluttes.begrunnelse,
                        null,
                    ),
                )
            }.onFailure { e ->
                log.warn("Kunne ikke avslutte oppfølging: {}", e.message)
            }.isSuccess
        }

        val avsluttedeBrukere = resultat.count { it }
        val ikkeAvsluttedeBrukere = resultat.count { !it }

        log.info("Avsluttet oppfølging for {} brukere", avsluttedeBrukere)
        log.info("Kunne ikke avslutte oppfølging for {} brukere", ikkeAvsluttedeBrukere)

        return AvsluttResultat(avsluttedeBrukere, ikkeAvsluttedeBrukere)
    }

    @PostMapping("/avsluttOppfolgingsperiode")
    fun avsluttOppfolgingsperiode(@RequestBody oppfolgingsperiodeSomSkalAvsluttes: AvsluttOppfolgingsperiodePayload): Boolean {
        sjekkTilgangTilAdmin()
        val innloggetBruker = authService.getInnloggetVeilederIdent()
        try {
            avsluttOppfolgingService.adminAvsluttSpesifikkOppfolgingsperiode(
                AktorId.of(oppfolgingsperiodeSomSkalAvsluttes.aktorId),
                innloggetBruker,
                oppfolgingsperiodeSomSkalAvsluttes.begrunnelse,
                oppfolgingsperiodeSomSkalAvsluttes.oppfolgingsperiodeUuid
            )
            return true
        } catch (e: Exception) {
            log.warn("Kunne ikke avslutte oppfølgingsperiode: {}", e.message)
            return false
        }
    }

    @PostMapping("/republiser/oppfolgingshendelse")
    fun republiserOppfolgingshendelse(@RequestBody aktorId: AktorId) {
        sjekkTilgangTilAdmin()
        republiserOppfolgingshendelseService.republiserOppfolgingshendelseForBruker(aktorId)
    }

    private fun sjekkTilgangTilAdmin() {
        if (!authService.erInternBruker()) throw ForbiddenException("Må være internbruker")
        authService.sjekkAtApplikasjonErIAllowList(listOf(POAO_ADMIN))
    }
}